src/modules/scraper/adapter/messaging/DepositReferralConsumer.ts

```diff
@@ -15,44 +17,36 @@ export class DepositReferralConsumer {
   constructor(
     @InjectRepository(Deposit) private depositRepository: Repository<Deposit>,
     private ethProvidersService: EthProvidersService,
+    private referralService: ReferralService,
+    private appConfig: AppConfig,
   ) {}
 
   @Process({ concurrency: 10 })
   private async process(job: Job<DepositReferralQueueMessage>) {
     const { depositId } = job.data;
     const deposit = await this.depositRepository.findOne({ where: { id: depositId } });
-    if (!deposit) throw new Error("Deposit not found");
+    if (!deposit) return;
     const { depositTxHash, sourceChainId } = deposit;
     const transaction = await this.ethProvidersService.getCachedTransaction(sourceChainId, depositTxHash);
+    const block = await this.ethProvidersService.getCachedBlock(sourceChainId, transaction.blockNumber);
+    const blockTimestamp = parseInt((new Date(block.date).getTime() / 1000).toFixed(0));
 
     if (!transaction) throw new Error("Transaction not found");
-    const referralAddress = this.extractReferralAddress(transaction.data);
-    await this.depositRepository.update({ id: deposit.id }, { referralAddress });
-  }
 
-  private extractReferralAddress(data: string) {
-    const coder = new ethers.utils.AbiCoder();
-    // strip hex method identifier
-    const dataNoMethod = ethers.utils.hexDataSlice(data, 4);
-    // keep method hex identifier
-    const methodHex = data.replace(dataNoMethod.replace("0x", ""), "");
-    const decodedData = coder.decode(
-      ["address", "address", "uint256", "uint256", "uint64", "uint32"],
-      ethers.utils.hexDataSlice(data, 4),
-    );
-    const encoded = coder.encode(["address", "address", "uint256", "uint256", "uint64", "uint32"], decodedData);
-    const fullEncoded = methodHex + encoded.replace("0x", "");
-    const referralData = data.replace(fullEncoded, "");
-    if (referralData.length >= 40) {
-      const potentialAddress = referralData.slice(referralData.length - 40);
-      try {
-        const address = ethers.utils.getAddress(`0x${potentialAddress}`);
-        return address;
-      } catch {
-        return undefined;
+    const { referralDelimiterStartTimestamp } = this.appConfig.values.app;
+    let referralAddress: string | undefined = undefined;
+
+    if (referralDelimiterStartTimestamp && blockTimestamp >= referralDelimiterStartTimestamp) {
+      referralAddress = this.referralService.extractReferralAddressUsingDelimiter(transaction.data);
+    } else {
+      referralAddress = this.referralService.extractReferralAddress(transaction.data);
+
+      if (referralAddress) {
+        const nonce = await this.ethProvidersService.getProvider(ChainIds.mainnet).getTransactionCount(referralAddress);
+        if (nonce === 0) referralAddress = undefined;
```

Comment:
also check the nonce if we found referral address prior to the configured block number

Reply:
+1